Dockerfile

```diff
@@ -1,8 +1,53 @@
-FROM python:3.6
-RUN set -x && \
-	pip install nornir
-VOLUME [ "/nornir" ]
-WORKDIR "/nornir"
-ENTRYPOINT [ "/usr/local/bin/python" ]
-#usage:
-# from inside your nornir base directory run: docker run -it --rm -v `pwd`:/nornir nornir:latest get_facts.py
+ARG PYTHON
+FROM python:${PYTHON}-slim-stretch AS builder
+
+WORKDIR /nornir
+ARG POETRY_PATH="/root/.poetry/bin/poetry"
+
+RUN apt-get update \
+    && apt-get install -yq curl git \
+    && curl -sSL https://raw.githubusercontent.com/sdispater/poetry/master/get-poetry.py | python \
+    && python -m venv .venv \
+    && ${POETRY_PATH} config settings.virtualenvs.in-project true
+
+COPY pyproject.toml .
+COPY poetry.lock .
+
+# Dependencies change more often, so we break RUN to cache the previous layer
+RUN ${POETRY_PATH} run pip install --no-cache-dir -U pip setuptools \
+    && ${POETRY_PATH} install --no-interaction
+
+# Source code changes even more often, so we cache another layer
+COPY nornir nornir
```

Comment:
to make better use of the cache you should invert the order, that way you'd save copying the README file because nornir folder changed.